### PR TITLE
Finish physical redesign

### PR DESCRIPTION
**Pull request by thejohnfreeman**
This is the final and most disruptive phase of the [physical redesign](https://github.com/XRPLF/XRPL-Standards/discussions/121). It makes these changes:

- CMake modules are moved into their own top-level `cmake/` directory.
- libxrpl headers are moved into `include/xrpl/`.
- libxrpl sources are moved to `src/libxrpl/`.
- rippled sources are moved to `src/xrpld/`.
- The `ripple` C++ namespace is unchanged.
- libxrpl headers are installed to _both_ the `xrpl/` and `ripple/` subdirectories of the header installation prefix. (`ripple/` is a symbolic link to `xrpl/`, even on Windows.) They include each other with path prefix `xrpl/`, but existing projects can continue to include them with path prefix `ripple/` and migrate to prefix `xrpl/` at their own pace.
- rippled is installed as _both_ `rippled` and `xrpld`. (`xrpld` is a symbolic link to `rippled`, even on Windows.)
- CMake source lists are computed from globs. Newly added source and header files are automatically built and installed without any intervention from contributors.
- CMake target names are cleaned up and standardized.

There is a non-trivial (but, I think, easy) process to merge existing branches-in-progress with this one. I will describe it here. Look at the [sequence of commits](https://github.com/XRPLF/rippled/pull/4997/commits). Notice the commit with subject "Add bin/physical.sh". Call that the **checkpoint** commit. There are no file movements before the checkpoint commit. The four commits following the checkpoint commit are generated by running the `bin/physical.sh` script added in the checkpoint commit.

When you want to merge, you should:

- Merge with the latest ancestor of this branch in `develop`. Resolve conflicts yourself. **Be sure that _all_ of the source files added in your branch (if any) appear in the correct source lists in `Builds/CMake/RippledCore.cmake`, between the `BEGIN`/`END` marker comments ([example](https://github.com/XRPLF/rippled/blob/2df635693d54b8c17c428faf6ef99da2f14a9382/Builds/CMake/RippledCore.cmake#L68-L198)) .** The restructure script moves only the files in these lists, and stops early if any files are leftover (because they were omitted from the lists). Remember the commit ID of this merge. Call it your **marker** commit.
    ```
    git merge c706926
    marker=$(git log -1 --pretty=format:%h)
    ```
- Merge with the checkpoint commit. Resolve conflicts yourself. (I will try to keep this instruction up-to-date, but the commit ID below may be incorrect. Please double-check the [commits](https://github.com/XRPLF/rippled/pull/4997/commits) in this PR to identify the correct checkpoint commit.)
    ```
    git merge 64a6c40
    ```
- Regenerate the next four commits by executing `bin/physical.sh`, passing it your marker commit from the first step:
    ```
    ./bin/physical.sh ${marker}
    ```
- Check the sort order in `Builds/levelization/results/ordering.txt`. Lines beginning with `xrpl.` must appear before lines beginning with `xrpld`. Different versions of `sort` have been inconsistent on this matter in my experience. On my machine, `sort` puts `xrpl.` after `xrpld`, which does not match the output of `sort` in our CI job that checks the file.
    ```
    ${EDITOR} Builds/levelization/results/ordering.txt
    git add --update .
    git commit --amend
    ```

Then you can `git merge -X ours $upstream` where `upstream` points to the head of this branch. It will automatically resolve conflicts in your branch's favor. Then you can `git diff $upstream` and see your work ported.

This PR requires the same finishing steps as #4966:

- Confirm it has at least two approvals from qualified reviewers.
- Rebase the commits on top of `develop`. The commit sequence that predates the review should be preserved. Fixes made since the review opened should be merged into earlier commits.
- Rewrite non-substantive (e.g. formatting) commits as authored by "Pretty Printer \<cpp@ripple.com>".
- Add rewritten non-substantive commit IDs to `.git-blame-ignore-revs`.
- Confirm the branch still passes all tests after making these changes.

### CODE REVIEW

**Comment by Bronek** [comment]
Note, this is moving files from `impl` and `details` to their parent directories, which I think is a problem because that expands the `libxrpl` API surface with the following files (which were designed to be implementation and not API)

```
json/impl/json_assert.h

protocol/impl/STVar.h  protocol/impl/b58_utils.h  protocol/impl/secp256k1.h  protocol/impl/token_errors.h

resource/impl/Entry.h  resource/impl/Import.h  resource/impl/Key.h  resource/impl/Kind.h  resource/impl/Logic.h  resource/impl/Tuning.h

server/impl/BaseHTTPPeer.h  server/impl/BaseWSPeer.h  server/impl/JSONRPCUtil.h  server/impl/PlainHTTPPeer.h  server/impl/SSLHTTPPeer.h  server/impl/ServerImpl.h
server/impl/BasePeer.h      server/impl/Door.h        server/impl/LowestLayer.h  server/impl/PlainWSPeer.h    server/impl/SSLWSPeer.h    server/impl/io_list.h
```

File: conanfile.py [hunk 1]
```diff
@@ -32,7 +32,6 @@ class Xrpl(ConanFile):
         'nudb/2.0.8',
         'openssl/1.1.1u',
         'protobuf/3.21.9',
-        'snappy/1.1.10',
```
[the remark this answers is not in the document]
Comment (thejohnfreeman) [reply conanfile.py:35]: We never link directly to Snappy, nor include its headers, nor do we need to resolve a version conflict.

**Comment by Bronek** [comment]
> Note, this is moving files from `impl` and `details` to their parent directories, which I think is a problem because that expands the `libxrpl` API surface with the following files (which were designed to be implementation and not API)
> 
> ```
> json/impl/json_assert.h
> 
> protocol/impl/STVar.h  protocol/impl/b58_utils.h  protocol/impl/secp256k1.h  protocol/impl/token_errors.h
> 
> resource/impl/Entry.h  resource/impl/Import.h  resource/impl/Key.h  resource/impl/Kind.h  resource/impl/Logic.h  resource/impl/Tuning.h
> 
> server/impl/BaseHTTPPeer.h  server/impl/BaseWSPeer.h  server/impl/JSONRPCUtil.h  server/impl/PlainHTTPPeer.h  server/impl/SSLHTTPPeer.h  server/impl/ServerImpl.h
> server/impl/BasePeer.h      server/impl/Door.h        server/impl/LowestLayer.h  server/impl/PlainWSPeer.h    server/impl/SSLWSPeer.h    server/impl/io_list.h
> ```

@thejohnfreeman thanks for addressing this concern.

**Review by Bronek — APPROVED** [review]
I like this a lot.

**Review by Bronek — COMMENTED** [review]
I noticed that the most recent rebase has reverted the change to which I refer here 
 https://github.com/XRPLF/rippled/pull/4997#issuecomment-2154849596


Is it intentional ?

**Comment by Bronek** [comment]
> I noticed that the most recent rebase has reverted the change to which I refer here [#4997 (comment)](https://github.com/XRPLF/rippled/pull/4997#issuecomment-2154849596)

Thank @thejohnfreeman , I see that's fixed now.

**Comment by scottschurr** [comment]
@thejohnfreeman, this is the branch I tried: https://github.com/scottschurr/rippled/tree/blocked-book

**Comment by thejohnfreeman** [comment]
I just merged that branch without a hitch: https://github.com/thejohnfreeman/rippled/tree/blocked-book

I'll try to replicate your experience on macOS soon.

**Comment by scottschurr** [comment]
Thanks.  And if there are any experiments you would like me to try I'm happy to do so.  Best of luck.

**Comment by thejohnfreeman** [comment]
The problem is what I expected, the environment. The requirements I found missing on macOS are:

- GNU `sed`
- GNU `find`
- Bash >= 4 (for `readarray`)
- `clang-format-10` must be on the `PATH`

Can you run the script in a Linux environment?

**Comment by scottschurr** [comment]
I just checked my Mac for the versions of some of the utilities the conversion script uses:
```
$ awk --version
awk version 20200816
$ bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin21)
Copyright (C) 2007 Free Software Foundation, Inc.
$ what /usr/bin/find
/usr/bin/find
	PROGRAM:find  PROJECT:shell_cmds-240.100.15
	PROGRAM:find  PROJECT:shell_cmds-240.100.15
$ what /usr/bin/sed
/usr/bin/sed
	PROGRAM:sed  PROJECT:text_cmds-138.100.3
	PROGRAM:sed  PROJECT:text_cmds-138.100.3
```
It sounds like the version of `bash` is just not gonna work with your script.  I'm less sure about the others.  Thoughts?

**Comment by thejohnfreeman** [comment]
On my Mac, I installed Bash from Homebrew. It has version 5.2. Mac's versions of `find` and `sed` come from FreeBSD (to my knowledge) with version strings that are inscrutable to me. The `awk` may be compatible.

As a reminder, I'm willing to manually merge anyone's in-progress branch on demand. They just need to resolve the conflicts up to whatever commit precedes the merge of this PR.

**Review by scottschurr — APPROVED** [review]
With @thejohnfreeman's guidance I was able to run the `bin/physical.sh` script successfully on my Mac with some minor edits.  First, using homebrew, I installed...

1. bash,
2. gawk,
3. gsed,
4. gfind, and
5. clang-format-11.

I then I edited `physical.sh` so it used the newly installed `bash`.  And I replaced the appropriate names in the shell script.

Once that was done the modified shell script ran perfectly for me.  I compared the results on my branch to the top of John's `restructure` branch and got the differences that I expected to see.  The restructured branch built, passed unit tests, synced, and stopped gracefully.

**Review by USDTASTRC555 — APPROVED** [review]

**Pull request merged**